crates/iceberg/src/expr/visitors/manifest_evaluator.rs

```diff
@@ -103,98 +106,245 @@ impl BoundPredicateVisitor for ManifestFilterVisitor<'_> {
+            let Literal::Primitive(PrimitiveLiteral::String(upper_bound)) = upper_bound else {
+                return Err(Error::new(
+                    ErrorKind::Unexpected,
+                    "Cannot perform starts_with on non-string upper bound",
+                ));
+            };
+
+            let min_len = upper_bound.chars().count().min(prefix_len);
+            let truncated_upper_bound = upper_bound.chars().take(min_len).collect::<String>();
+            if prefix > &truncated_upper_bound {
+                return ROWS_CANNOT_MATCH;
+            }
+        }
+
+        ROWS_MIGHT_MATCH
     }
 
     fn not_starts_with(
         &mut self,
         reference: &BoundReference,
-        literal: &Datum,
+        datum: &Datum,
         _predicate: &BoundPredicate,
     ) -> crate::Result<bool> {
-        todo!()
+        let field: &FieldSummary = self.field_summary_for_reference(reference);
+
+        if field.contains_null || field.lower_bound.is_none() || field.upper_bound.is_none() {
+            return ROWS_MIGHT_MATCH;
+        }
+
+        let PrimitiveLiteral::String(prefix) = datum.literal() else {
+            return Err(Error::new(
+                ErrorKind::Unexpected,
+                "Cannot perform not_starts_with on non-string value",
+            ));
+        };
+
+        let prefix_len = prefix.chars().count();
+
+        // not_starts_with will match unless all values must start with the prefix. This happens when
+        // the lower and upper bounds both start with the prefix.
+        if let Some(lower_bound) = &field.lower_bound {
+            let Literal::Primitive(PrimitiveLiteral::String(lower_bound)) = lower_bound else {
+                return Err(Error::new(
+                    ErrorKind::Unexpected,
+                    "Cannot perform not_starts_with on non-string lower bound",
+                ));
+            };
+
+            // if lower is shorter than the prefix then lower doesn't start with the prefix
+            if prefix_len > lower_bound.chars().count() {
+                return ROWS_MIGHT_MATCH;
+            }
+
+            let truncated_lower_bound = lower_bound.chars().take(prefix_len).collect::<String>();
+            if prefix == &truncated_lower_bound {
+                if let Some(upper_bound) = &field.upper_bound {
+                    let Literal::Primitive(PrimitiveLiteral::String(upper_bound)) = upper_bound
+                    else {
+                        return Err(Error::new(
+                            ErrorKind::Unexpected,
+                            "Cannot perform not_starts_with on non-string upper bound",
+                        ));
+                    };
+
+                    // if upper is shorter than the prefix then upper can't start with the prefix
+                    if prefix_len > upper_bound.chars().count() {
+                        return ROWS_MIGHT_MATCH;
+                    }
+
+                    let truncated_upper_bound =
```

Comment:
I think we can avoid the extra String allocation by using `char_indices()` to get the prefix_len and then use a slice for comparison `let truncated_upper_bound = &upper_bound[..prefix_len];` // haven't tested it though

Reply:
I think @marvinlankenke's suggestion is indeed a better one than mine here

Reply:
On reflection, I think we can just do this:

```rust
if prefix.as_bytes().eq(&lower_bound.as_bytes()[..prefix.len()]) {
```

Since we know `prefix` is correctly formed UTF-8 as it is a `&String`, if all we are doing is comparing equality of the first n chars, if all of the initial bytes of the bound are the same as in `prefix` then the slice must also fall on a char boundary.